Remove dead code in installer

haampie opened the pull request:
The `InstallDirectoryAlreadyExistsError` exception is never raised since the code path only ever happens when the install directory does not exist.

See https://github.com/spack/spack/blob/develop/lib/spack/spack/installer.py#L1294-L1297, this the only call site of create_install_directory, except for the tests in lib/spack/spack/test/directory_layout.py, but they use the function as a helper where the install prefix indeed does not exist.

haampie committed:
Remove tests that don't really test anything (they pass with previous commit)

haampie committed:
Remove more dead code

tldahlgren commented:
I'm curious.  How did you determine this code is dead?

My concern about removing it would be if someone attempts concurrent (or overlapping) builds on a system that does not support posix locks.

@tgamblin Any thoughts on whether this code should be considered dead?

haampie commented:
Hi Tamara, I've tried to add descriptive commit messages, but github doesn't do a great job at showing these past the first line, you'd have to click the first one.

This is how it's called:

https://github.com/spack/spack/blob/bdb02ed535118b6e270d07aa718a04b943c6c71e/lib/spack/spack/installer.py#L1300-L1303

https://github.com/spack/spack/blob/bdb02ed535118b6e270d07aa718a04b943c6c71e/lib/spack/spack/directory_layout.py#L320-L325

https://github.com/spack/spack/blob/bdb02ed535118b6e270d07aa718a04b943c6c71e/lib/spack/spack/directory_layout.py#L342-L348

And this function is never called elsewhere except for tests:

```
$ grep create_install_directory --include *.py --exclude-dir=test -r ~/spack/lib/
/home/harmen/spack/lib/spack/spack/installer.py:            spack.store.layout.create_install_directory(pkg.spec)
/home/harmen/spack/lib/spack/spack/directory_layout.py:    def create_install_directory(self, spec):
/home/harmen/spack/lib/spack/spack/directory_layout.py:    def create_install_directory(self, spec):
/home/harmen/spack/lib/spack/spack/directory_layout.py:    """Raised when create_install_directory is called unnecessarily."""
```

So, it makes no sense to call `check_installed`, and then it makes no sense to have that exception, and then it makes no sense to handle that exception.

haampie committed:
Don't run check_installed when creating new prefix

`check_installed` should not be run when creating a new prefix, because:

a) the prefix does not exist (it's only ever run in a code path that
first destroys it if it exists)
b) even if it would exist, it does a lot of validation of the spec.yaml
file to ensure there is no prefix collision, but this can't happen as
the database guarantees this.

tgamblin commented:
@haampie: Trying desperately to get https://github.com/spack/spack/pull/22845 merged right now.  So let's rebase and see what's left after that goes in.

haampie commented:
Thanks, turns out it's not a conflicting pr, I just successfully rebased it on that locally

scheibelp commented:
Thanks!
